### ISSUE TITLE
Are tables implemented?

### ISSUE THREAD

**Issue by deanm0000** [vinci1it2000/formulas#105]
I got this error:

FormulaError: ('Not a valid formula:\n%s', '=Table6[[#This Row],[MWh]]/Table6[[#This Row],[MW]]')

I'm guessing Tables aren't implemented and that's why this doesn't work since it's just division.  Is that right?

On that note, is there a way to exclude tabs when loading a file via `formulas.ExcelModel().loads(fpath).finish()` or to warn for formula errors instead of erroring out?

**Comment by vinci1it2000** [comment]
Yes tables are not implemented. Can you share the reference and definition of tables? So it can be in the pipeline of futures development

**Comment by deanm0000** [comment]
Thanks for the follow up.

Is this what you mean by reference?
https://support.microsoft.com/en-us/office/overview-of-excel-tables-7ab0bb7d-3a9e-4b56-a3c9-6c94334e492c

Also, I'm still wondering if there's a way to exclude tabs or warn about errors instead of raising an exception.

**Comment by FynnFreyer** [comment]
Hi, 

I'm working on a project that needs this functionality as well, and my manager told me that long term it would be much cleaner to use this library, instead of using our current way of `libreoffice --headless --convert-to csv [...]`. So, when I have capacity, I would like to work on implementing tables if you'd be willing to accept a PR regarding this @vinci1it2000.

Anyway, I'm putting some loose thoughts into this issue, in case this might be helpful down the line:

An easily digestible overview of the syntax of structured references (e.g. `=<table_name>[[<column_1>],[<column_2>]]`) can be found [here](https://support.microsoft.com/en-us/office/using-structured-references-with-excel-tables-f5ed2452-2337-4f71-bed3-c8ae6d2b276e) in the "Structured reference syntax rules" section and following.

I had a (very) quick (and cursory) glance around the code base, so I don't understand it that well, but it seems that adding another token filter to the `Parser` class would be the way to do this. There could be more `Operand` classes for tables, column specifiers, item specifiers, and table specifiers and another `Operator` for referencing with brackets. Are `Operator`s designed with something other than the typical infix notation in mind? If not, it might make sense to make this a special kind of `Function` instead? As I said, it was a very cursory glance, and these thoughts are preliminary.

Is there any documentation available on what exactly those `ast` and `process` methods need to do? (I mean, `ast` probably builds up an abstract syntax tree, and `process` might calculate things (?), but that's really just an initial guess, because I don't want to read the code and figure it out atm.)

Fortunately, `openpyxl`s `Worksheet`s already give easy access to a dictionary of tables.

This is probably a case of YAGNI, but it might make sense to consider [references to other worksheets](https://support.microsoft.com/en-us/office/create-or-change-a-cell-reference-c7b8b95d-c594-4488-947e-c835903cebaa) (section "Create a cell reference to another worksheet", e.g. `=<sheet_name>!A1:B1`), or [named ranges](https://support.microsoft.com/en-us/office/define-and-use-names-in-formulas-4d0f13ac-53b7-422e-afd2-abd7ff379c64) (accessible in `openpyxl` via `defined_names`, maybe somewhat related to #114) when implementing structured references for tables.
Maybe `formulas` can already do those, then nvm.

Also, because Excel-developers seem to be masochists (or sadists), you can actually [reference data from different workbooks](https://support.microsoft.com/en-us/office/create-an-external-reference-link-to-a-cell-range-in-another-workbook-c98d1803-dd75-4668-ac6a-d7cca2a9b95f), but I would think that's super out of scope.

Besides, the people who use this "feature" deserve the hell they are wading through. I mean modularization is nice and everything, but this is just sick. /s

**Comment by dberardo-com** [comment]
any update on this issue? would it be possible to define tables as a custom formula or would that be to hard ?

@FynnFreyer ?

**Comment by FynnFreyer** [comment]
Hi @vinci1it2000 I wrote you an email

**Comment by FynnFreyer** [comment]
@dberardo-com, thanks for pinging me, because I totally missed the first mention. I have not been working on this as yet, and if you need something soon, I'd suggest you take a look at [exporting to csv](https://help.libreoffice.org/7.4/en-GB/text/shared/guide/csv_params.html?DbPAR=SHARED) via LibreOffice. Maybe that helps you for your use case (it's the ugly hack I'm using in another project rn)

Probably not a good idea to hold your breath and wait for the solution here atm, since I can't give any sort of eta as to when I get around to working on this ^^'

**Comment by TAVMWB** [comment]
I ran in to the same issue, and tried the exporting to csv with libreoffice as @FynnFreyer referred to but I couldnt get that to work.
Instead manually corrected the table references to cell references with an Excel macro in the file before loading it with formulas: 

```vbscript
Sub ConvertTablesToRange()
    Dim wks As Worksheet, objList As ListObject
    Set wks = ActiveWorkbook.ActiveSheet
    For Each objList In wks.ListObjects
        objList.Unlist
    Next objList
End Sub
```

After running this script I could load the excel sheet in formulas without any problems. It would be nice if the package would support tables natively though of course.
